### PR TITLE
Refactor GetPath in JsonPatchDocumentOfT

### PR DESCRIPTION
**Pull request by jbagga**
Addresses #98 

cc @kichalla

### CODE REVIEW

File: test/Microsoft.AspNetCore.JsonPatch.Test/JsonPatchDocumentJsonPropertyAttributeTest.cs [hunk 1]
```diff
@@ -12,6 +13,40 @@ namespace Microsoft.AspNetCore.JsonPatch
     public class JsonPatchDocumentJsonPropertyAttributeTest
     {
         [Fact]
+        public void Add_ToRoot_OfListOfObjects()
+        {
+            var patchDoc = new JsonPatchDocument<List<JsonPropertyObject>>();
+            patchDoc.Add(p => p, new JsonPropertyObject());
+
+            var serialized = JsonConvert.SerializeObject(patchDoc);
+            // serialized value should have "AnotherName" as path
```
Comment (kichalla) [inline test/Microsoft.AspNetCore.JsonPatch.Test/JsonPatchDocumentJsonPropertyAttributeTest.cs]: this comment is not valid here...probably a copy-paste error

File: test/Microsoft.AspNetCore.JsonPatch.Test/JsonPatchDocumentJsonPropertyAttributeTest.cs [hunk 2]
```diff
@@ -12,6 +13,40 @@ namespace Microsoft.AspNetCore.JsonPatch
...
+            // get path
+            var pathToCheck = deserialized.Operations.First().path;
+            Assert.Equal("/-", pathToCheck);
+        }
+
+        [Fact]
+        public void Add_ToRoot_OfAnObject()
+        {
+            var patchDoc = new JsonPatchDocument<JsonPropertyObject>();
+            patchDoc.Add(p => p, new JsonPropertyObject());
```
Comment (kichalla) [inline test/Microsoft.AspNetCore.JsonPatch.Test/JsonPatchDocumentJsonPropertyAttributeTest.cs]: This test doesn't make much sense here. (Adding to a POCO object itself). Also can Path be empty? i suppose that's an error case.

File: src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs [hunk 3]
```diff
@@ -717,7 +717,13 @@ IList<Operation> IJsonPatchDocument.GetOperations()
...
         private string GetPath<TProp>(Expression<Func<TModel, TProp>> expr)
         {
-            return "/" + GetPath(expr.Body, true).ToLowerInvariant();
+            var path = GetPath(expr.Body, true);
+            if (!String.IsNullOrEmpty(path))
+            {
+                path = "/" + path;
+            }
+
+            return path.ToLowerInvariant();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: Should this change just go into the `case ExpressionType.Parameter` block?

Without some examples I'm a little skeptical that this fixes all of the cases we care about.

Comment (jbagga) [reply src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: Can we discuss this? I am not sure how I can preemptively remove the extra slash in the `case ExpressionType.Parameter` block?

**Comment by jbagga** [comment]
@kichalla Added unit tests for GetPath

**Comment by jbagga** [comment]
@rynowak @kichalla Updated

File: src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs [hunk 4]
```diff
@@ -776,33 +772,24 @@ private string GetPropertyNameFromMemberExpression(MemberExpression memberExpres
...
+                || expressionType == ExpressionType.Call
+                || expressionType == ExpressionType.Convert
+                || expressionType == ExpressionType.MemberAccess);
+
         }
 
-        private static string GetIndexerInvocation(Expression expression)
+        // Evaluates the value of the key or index which may be an int or a string, 
+        // or some other expression type.
+        // The expression is converted to executable code and the resulting delegate is returned as a string.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: `The expression is converted to a delegate and the result of executing the delegate is returned as a string`

File: src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs [hunk 5]
```diff
@@ -705,61 +705,57 @@ IList<Operation> IJsonPatchDocument.GetOperations()
...
+                {
+                    return path.ToLowerInvariant();
+                }
+            }
+
+            return "/" + path.ToLowerInvariant();
         }
 
-        private string GetPath(Expression expr, bool firstTime)
+        private List<string> GetPath(Expression expr, bool firstTime)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: I think the `first time` parameter can be removed

Comment (rynowak) [inline src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: It wouldn't hurt to rename this to something like `GetPathSegments` its usage isn't really interchangable with `GetPath` that returns a string

File: test/Microsoft.AspNetCore.JsonPatch.Test/GetPathTest.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Xunit;
+
+namespace Microsoft.AspNetCore.JsonPatch
+{
+    public class GetPathTest
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.JsonPatch.Test/GetPathTest.cs]: This should be called `JsonPatchDocumentGetPathTest`.

File: src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs [hunk 7]
```diff
@@ -705,61 +705,57 @@ IList<Operation> IJsonPatchDocument.GetOperations()
...
+            var segments = GetPathSegments(expr.Body);
+            var path = String.Join("/", segments);
+            if (position != null)
+            {
+                path += "/" + position;
+                if (segments.Count == 0)
+                {
+                    return path.ToLowerInvariant();
+                }
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: what about a case where `segments.Count == 0` and `position != null`? It looks like the value would be missing the leading slash.

Comment (rynowak) [reply src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: A good way to think about this is that `path = "/" + string.Join("/", segments)` will always generate the right thing even when segments is empty.

Then you just have to deal with appending the position.

Comment (jbagga) [reply src/Microsoft.AspNetCore.JsonPatch/JsonPatchDocumentOfT.cs]: This case is covered. Adding tests where position is not null

**Comment by jbagga** [comment]
@rynowak @kichalla Added non-null positions tests
